moabb/contexts/evaluations.py

```diff
@@ -0,0 +1,190 @@
+        ep = concatenate_epochs(all_events)
+        # previously multipled data by 1e6
+        X, y = (ep.get_data(), ep.events[:, -1])
+        return X, y
+    
+    def score(self, clf, X, y, scoring):
+        cv = KFold(5, shuffle=True, random_state=self.random_state)
+
+        y_uniq = np.unique(y)
+        if len(y_uniq) == 2:
+            for ind, l in enumerate(y_uniq):
+                y[y==l] = ind
+        acc = cross_val_score(clf, X, y, cv=cv,
+                              scoring=scoring, n_jobs=self.n_jobs)
+        return acc.mean()
+
+class CrossSubjectEvaluation(TrainTestEvaluation):
+    """Cross Subject evaluation Context.
+
+    Evaluate performance of the pipeline trained on all subjects but one,
+    concatenating sessions.
+
+    Parameters
+    ----------
+    datasets : List of Dataset instances.
+        List of dataset instances on which the pipelines will be evaluated.
+    pipelines : Dict of pipelines instances.
+        Dictionary of pipelines. Keys identifies pipeline names, and values
+        are scikit-learn pipelines instances.
+
+    See Also
+    --------
+    BaseContext
+    """
+    def evaluate(self, dataset, subject, clf, paradigm):
+        """Prepare data for classification."""
+        # requires that subject be an int
+        s = subject-1
+        self.ind_cache[s] = self.ind_cache[s]*0
+        allX = np.concatenate(self.X_cache)
+        ally = np.concatenate(self.y_cache)
+        groups = np.concatenate(self.ind_cache)
+        # re-generate s group label
+        self.ind_cache[s] = np.ones(self.ind_cache[s].shape)
+        t_start = time()
+        score = self.score(clf, allX, ally, groups, paradigm.scoring)
+        duration = time() - t_start
+        return {'time': duration, 'dataset': dataset.code, 'id': subject, 'score': score, 'n_samples': len(ally)}
+
+    def preprocess_data(self, d, paradigm):
+        assert len(d.subject_list) > 1, "Dataset {} has only one subject".format(d.code)
+        self.X_cache = []
+        event_id = d.selected_events
+        if not event_id:
+            raise(ValueError("Dataset had no selected events"))
+        self.y_cache = []
+        self.ind_cache = []
+        for s in d.subject_list:
+            sub = d.get_data([s], stack_sessions=True)[0]
+            # get all epochs for individual files in given subject
+            epochs = paradigm._epochs(sub, event_id, d.interval)
+            # equalize events from different classes
+            X, y = self.extract_data_from_cont(epochs, event_id)
+            self.X_cache.append(X)
+            self.y_cache.append(y)
+            self.ind_cache.append(np.ones(y.shape))
+        
+    def score(self, clf, X, y, groups, scoring):
+        y_uniq = np.unique(y)
+        if len(y_uniq) == 2:
+            for ind, l in enumerate(y_uniq):
+                y[y==l] = ind
+        acc = cross_val_score(clf, X, y, cv=[(np.nonzero(groups==1)[0], np.nonzero(groups==0)[0])],
```

Comment:
any reason not using [LeaveOneGroupOut](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.LeaveOneGroupOut.html) or a similar class.

if there is only one partition, we might want to just use fit() and score()

Reply:
The way I wrote it, all evaluations work like this in pseudocode:
```
pre_process_dataset(d)
for subject in dataset:
    process_subject(subject)
```
which means that we can't do LeaveOneGroupOut since that doesn't generate a single value per subject, but rather a single value per dataset. It's a bit convoluted, I admit...not sure if the best way to deal with it is how I have, or if I should simply reimplement the 'process' method here

Reply:
so you effectively doing a leave one subject out, just at a higer level.
this looks good to me.